Sources/Networking/OAuth/OAuthAuthenticator.swift

```diff
@@ -0,0 +1,133 @@
+        return response.statusCode == 401
+    }
+
+    public func isRequest(_ urlRequest: URLRequest, authenticatedWith credential: Credential) -> Bool {
+        let bearerToken = HTTPHeader.authorization(bearerToken: credential.accessToken).value
+        return urlRequest.headers["Authorization"] == bearerToken
+    }
+}
+
+// MARK: - Extensions -
+
+// MARK: - Token refreshing
+
+public extension OAuthAuthenticator {
+
+    static func refresh(
+        userCredential: OAuthUserCredential,
+        for session: Session,
+        completion: @escaping (Result<Credential, Error>) -> Void
+    ) {
+        // Set the proper user model on AFDataResponse or modify the example one
+        let requestHandler: (AFDataResponse<UserAuthenticatedModel>) -> Void = { response in
+            switch response.result {
+            case .success(let value):
+                Self.handle(userTokenRefreshUsing: value, completion: completion)
+            case .failure(let error):
+                Self.handle(tokenRefreshFailureUsing: error, completion: completion)
+            }
+        }
+
+        // Request the token refresh using the provided session, and call the requestHandler for the completionHandler on responseDecodable
+        // Don't forget to validate the request
+    }
+
+    static func refresh(
+        clientCredential: OAuthClientCredential,
+        for session: Session,
+        completion: @escaping (Result<Credential, Error>) -> Void
+    ) {
+        // Set the proper client model on AFDataResponse or modify the example one
+        let requestHandler: (AFDataResponse<ClientAuthenticatedModel>) -> Void = { response in
+            switch response.result {
+            case .success(let value):
+                Self.handle(clientTokenRefreshUsing: value, completion: completion)
+            case .failure(let error):
+                Self.handle(tokenRefreshFailureUsing: error, completion: completion)
+            }
+        }
+
+        // Request the client token refresh using the provided session, and call the requestHandler for the completionHandler on responseDecodable
+        // Don't forget to validate the request
+    }
+}
+
+// MARK: - Refresh handlers
+
+private extension OAuthAuthenticator {
+
+    // Modify the UserAuthenticatedModel to suit your needs
+    static func handle(
+        userTokenRefreshUsing model: UserAuthenticatedModel,
+        completion: @escaping (Result<Credential, Error>) -> Void
+    ) {
+        let credential = OAuthUserCredential(
+            accessToken: model.accessToken,
+            refreshToken: model.refreshToken,
+            expiration: Date(timeIntervalSinceNow: model.expiresIn)
+        )
+        completion(.success(.user(credential)))
+
+        // Notify and save the new model/required properties in sessionManager
+        let object = ["AccessTokenRefreshed": model]
+        NotificationCenter.default.post(name: .init(rawValue: "AccessTokenRefreshed"), object: object)
```

Comment:
Potentially can put these notification names to extensions as 
```
extension Notification.Name {
    
    static var accessTokenRefreshed: Notification.Name {
        .init("AccessTokenRefreshed")
    }
    
}
```

and use as `NotificationCenter.default.post(name: .accessTokenRefreshed, object: object)`

Same can be then done for bottom two (ClientTokenRefreshed, SessionExpired)